consensus/istanbul/backend/engine.go

```diff
@@ -921,10 +921,52 @@ func (sb *backend) snapshot(chain consensus.ChainReader, number uint64, hash com
 		logger.Trace("Stored voting snapshot to disk", "number", snap.Number, "hash", snap.Hash)
 	}
 
+	sb.regen(chain, headers)
+
 	sb.recents.Add(snap.Hash, snap)
 	return snap, err
 }
 
+// regen commits snapshot data to database
+func (sb *backend) regen(chain consensus.ChainReader, headers []*types.Header) {
+	if !sb.isRestoring.Load() && len(headers) > 1 {
```

Comment:
The loop condition check might be helpful in reducing a useless loop.
```
last := headers[len(headers)-1].Number
if last - (last % CheckpointInterval) >= headers[0].Number {
    // Start loop
}
```

Reply:
Good suggestion. Let me gather further opinion for this. What do you @ian0371  think about this shortcut?

Reply:
LGTM. Could you add the condition to the caller side?

```
// if checkpoints are in `headers`, store their snapshots
if last - (last % CheckpointInterval) >= headers[0].Number {
    sb.regen(chain, headers)
}
```

Reply:
How about to add into inner function? If not, the caller seems to get dirty like
```
if len(headers) > 1 {  
    from := headers[0].Number
    to := headers[len(headers)-1].Number

    if to.Uint64()-(to.Uint64()%uint64(params.CheckpointInterval)) >= from.Uint64() {
        sb.regen(chain, headers)
    }
}
```

versus (please put a special attention to the comparator)
```
if to.Uint64()-(to.Uint64()%uint64(params.CheckpointInterval)) < from.Uint64() {
    return
}
```


Of course, the place of guard in call side seems to stand out for readers, but let me ask where it's okay to go with.

Reply:
It's a bit annoying to get the `from` and `to` values redundantly, but it seems better to do it on the caller side since this is an execution condition for the `regen` itself. It also helps not to call `regen` in most cases (`len(headers) < 2`)

Reply:
Then please add a comment above the function implementation indicating when this function becomes active.

Reply:
Added. PTAL.